src/Stratis.Bitcoin/FinalizedBlockInfoRepository.cs

```diff
@@ -0,0 +1,202 @@
+    /// </remarks>
+    public interface IFinalizedBlockInfoRepository : IDisposable
+    {
+        /// <summary>Gets the finalized block hash and height.</summary>
+        /// <returns>Hash and height of a block that can't be reorged away from.</returns>
+        HashHeightPair GetFinalizedBlockInfo();
+
+        /// <summary>Loads the finalised block hash and height from the database.</summary>
+        Task LoadFinalizedBlockInfoAsync(Network network);
+
+        /// <summary>Saves the finalized block hash and height to the database if height is greater than the previous value.</summary>
+        /// <param name="hash">Block hash.</param>
+        /// <param name="height">Block height.</param>
+        /// <returns><c>true</c> if new value was set, <c>false</c> if <paramref name="height"/> is lower or equal than current value.</returns>
+        bool SaveFinalizedBlockHashAndHeight(uint256 hash, int height);
+    }
+
+    public class FinalizedBlockInfoRepository : IFinalizedBlockInfoRepository
+    {
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        /// <summary>Access to DBreeze database.</summary>
+        private readonly DBreezeEngine dbreeze;
+
+        /// <summary>Database key under which the block height of the last finalized block height is stored.</summary>
+        private static readonly byte[] finalizedBlockKey = new byte[0];
+
+        /// <summary>Height and hash of a block that can't be reorged away from.</summary>
+        private HashHeightPair finalizedBlockInfo;
+
+        /// <summary>Queue of finalized infos to save.</summary>
+        /// <remarks>All access should be protected by <see cref="queueLock"/></remarks>
+        private readonly Queue<HashHeightPair> finalizedBlockInfosToSave;
+
+        /// <summary>Protects access to <see cref="finalizedBlockInfosToSave"/>.</summary>
+        private readonly object queueLock;
+
+        /// <summary>Task that continously persists finalized block info to the database.</summary>
+        private readonly Task finalizedBlockInfoPersistingTask;
+
+        private readonly CancellationTokenSource cancellation;
+
+        private readonly AsyncManualResetEvent queueUpdatedEvent;
+
+        public FinalizedBlockInfoRepository(string folder, ILoggerFactory loggerFactory)
+        {
+            Guard.NotEmpty(folder, nameof(folder));
+            Guard.NotNull(loggerFactory, nameof(loggerFactory));
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+
+            Directory.CreateDirectory(folder);
+            this.dbreeze = new DBreezeEngine(folder);
+
+            this.finalizedBlockInfosToSave = new Queue<HashHeightPair>();
+            this.queueLock = new object();
+
+            this.queueUpdatedEvent = new AsyncManualResetEvent(false);
+            this.cancellation = new CancellationTokenSource();
```

Comment:
No, it's better to have it's own source that is cancelled on disposal. Otherwise we might have problems with testing (would be kinda weird if the component hangs on disposal because inner loop is waiting for nodeLifetime's token)

Reply:
Ah yeah makes sense! 👍